File: src/main/java/io/eliez/fintools/bin2brand/PrefixRangeParser.java
package io.eliez.fintools.bin2brand;

import lombok.experimental.UtilityClass;
import lombok.val;

@UtilityClass
public class PrefixRangeParser {

    static PrefixRange parse(String rangeSpec, RangePattern rangePattern) {
        val matcher = rangePattern.pattern().matcher(rangeSpec);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("invalid prefix spec (pattern): " + rangeSpec);
        }
        if (matcher.group(3) == null) {
            final String element = matcher.group(1);
            return new PrefixRange(element, element);
        } else {
            final String lower = matcher.group(1);
            final String upper = matcher.group(3);
            return new PrefixRange(lower, upper);
        }
    }

    static PrefixRange parseNumeric(String rangeSpec) {
        return parse(rangeSpec, RangePattern.NUMERIC);
    }
}

File: src/main/kotlin/io/eliez/fintools/bin2brand/PrefixRangeParser.kt
package io.eliez.fintools.bin2brand

class PrefixRangeParser {
    companion object {

        internal fun parse(rangeSpec: String, rangePattern: RangePattern): PrefixRange {
            val matcher = rangePattern.pattern.matcher(rangeSpec)
            if (!matcher.matches()) {
                throw IllegalArgumentException("invalid prefix spec (pattern): " + rangeSpec)
            }
            if (matcher.group(3) == null) {
                val element = matcher.group(1)
                return PrefixRange(element, element)
            } else {
                val lower = matcher.group(1)
                val upper = matcher.group(3)
                return PrefixRange(lower, upper)
            }
        }

        @JvmStatic
        fun parseNumeric(rangeSpec: String): PrefixRange {
            return parse(rangeSpec, RangePattern.NUMERIC)
        }
    }
}
